New share provider: Share by mail

schiessle opened the pull request:
This introduces a new share provider called "Share by mail" which allows you to enter a email address to the share dialog and send a personalized link.

![share-by-mail](https://cloud.githubusercontent.com/assets/1589737/17250083/10f52de6-55a3-11e6-8d88-98b8f15ad429.png)

This is work in progress, but at least the auto-complete function works already.

Fix #1703

schiessle committed:
introduce share by mail, ui part

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
add share by mail share provider

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle commented:
@jancborchardt I'm sure that's also interesting to you :wink:

nickvergessen commented:
missing from shipped.json

rullzer commented:
While I think the idea is good. The UX there is terrible. Basically it should fall back to sending the e-mail if no remote is found.

icewind1991 commented on apps/sharebymail/lib/FederatedShareProvider.php:
```diff
@@ -0,0 +1,528 @@
+		/*
+		 * Check if file is not already shared with the remote user
+		 */
+		$alreadyShared = $this->getSharedWith($shareWith, self::SHARE_TYPE_REMOTE, $share->getNode(), 1, 0);
```
`SHARE_TYPE_EMAIL`?

icewind1991 commented on apps/sharebymail/lib/FederatedShareProvider.php:
```diff
@@ -0,0 +1,528 @@
+		$currentServer = $this->addressHandler->generateRemoteURL();
+		$currentUser = $sharedBy;
+		if ($this->addressHandler->compareAddresses($user, $remote, $currentUser, $currentServer)) {
+			$message = 'Not allowed to create a federated share with the same user.';
```
email share, not federated share

icewind1991 commented:
The share provider looks like a lot of duplicate code from the other share providers, probably worth it having a go at factoring out the common logic

jancborchardt commented:
@schiessle awesome! :) Great that you’re pushing this!

I tend to agree with @rullzer – it’s a bit strange at the moment that you need to decide between remote and email. It should check automatically if a remote by that ID exists – if not share via email.

Also, can we tune that dropdown so that to the left of the remote user it shows their avatar (or colored circle with initial) and next to email it shows an email icon? Or do we need to do a major rewire for that?

schiessle commented:
> The share provider looks like a lot of duplicate code from the other share providers, probably worth it having a go at factoring out the common logic

yes, that's mainly copy&pased from the federated share provider... This will change a lot, you can ignore it for now

schiessle commented:
> I tend to agree with @rullzer – it’s a bit strange at the moment that you need to decide between remote and email. It should check automatically if a remote by that ID exists – if not share via email.

I'm not sure. This would mean that you would first hand the share to the federated share provider, which would try to create a share. If it fails it would return to the share manager which would then try to do a mail share. Such a behavior is also completely in-transparent to the user. I tend to let the user explicitly decide what he want to do. In case we find a exact match of a federated share or a email address we can hide the other option because it is really unlikely that a user has the exact same email address and federated cloud id. But if we enter something for the first time I would just let the user chose.

> Also, can we tune that dropdown so that to the left of the remote user it shows their avatar (or colored circle with initial) and next to email it shows an email icon? Or do we need to do a major rewire for that?

We don't have a avatar if your share with a user for the first time... Let's keep this for later. This is out of scope for this PR.

jancborchardt commented:
> I tend to let the user explicitly decide what he want to do.

Nextcloud tends to do stuff automatically. ;) If you share something with me via email – either cause my server is not available or you don’t have my federated cloud ID – it could be added to my Nextcloud as a federated share and communicated back to your server like that. Just automatically, there should be no difference and there doesn’t need to be, because in both cases you have the control over if I »can edit« or not.

mar1u5 commented:
Uhmm ... Isnt this covered by https://github.com/owncloud/core/25576

schiessle commented:
> Nextcloud tends to do stuff automatically. ;) If you share something with me via email – either cause my server is not available or you don’t have my federated cloud ID – it could be added to my Nextcloud as a federated share and communicated back to your server like that. 

That's confusing. Sounds to me like you mixed federated shares and mail shares up. :wink: 

Let me try to illustrate the two possible scenarios if we try to make it automatically:

1. We first try to do a federated share. If it succeed fine. But what do we do if it don't succeed? If we know it is a federated share ID we can retry it and the recipient will receive his share once his Nextcloud is back online -> everything is fine. If we try to be more intelligent as the user we would fall back to a mail share, we would send the mail to the cloud ID. The delivery will fail because in most cases the cloud ID is not a email address at the same time. We will have no chance to detect that sending the mail failed. The user who send the share will see it in his share list and assume that everything is fine because he entered a federated cloud id. -> this is a problem you can't solve.

2. So we could try to do it the other way around, first try to send a mail.  If it is a valid mail address everything fine. If the rare case happens (0.1%) that a user has the same federated cloud ID as email address this would be bad because he would never receive a federated share but always a email. But it is just 0.1% so let's ignore it for now. Let's assume the ID was a cloud ID, we tried to send a mail and will not have any response if the mail was delivered successfully or not, so we never know if we should retry a federated share -> solution is not feasible at all.

So we have two solutions to automate it and both don't work.

schiessle commented:
> Uhmm ... Isnt this covered by https://github.com/owncloud/core/25576

@mar1u5 link doesn't work

mar1u5 commented:
Was updated 1 minute later :unamused: ... The /pull/ was missing :wink:

schiessle commented:
As you said, it is something different. I don't want to have a way to share a public link by mail, this is already today possible. But to create multiple public links, one for each recipient transparent in the share dialog

jancborchardt commented:
cc @Rocco83 because this is the »generate unique link per email address so revocation on a per-person basis is possible« feature you are interested in. :)

schiessle committed:
update permissions

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

jancborchardt commented:
@schiessle any progress on this? Do you need any help or reviews? :)

schiessle committed:
add method to check if a share provider for a given type is loaded

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
handle case if no share-by-mail share provider is loaded

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
improve search for federated cloud ids and email adresses

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
show correct display name if we have the user in one of our address books

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
unified way to display remote shares and mail shares

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
if we have a exact match, either for the federated cloud id or for the
email address we only return the exact match. It is highly unlikely
that the exact same email address and federated cloud id exists

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
send mail for share-by-mail shares

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
don't show 'notify by mail' option or permissions not available for mail shares

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle commented:
I will look if I can add some unit tests, but beside that it is ready to test and review

schiessle committed:
find and show share-by mail links

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
unit tests updated and new added

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
enabled by default and add to shipped.json

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

nickvergessen commented on .gitignore:
```diff
@@ -16,6 +16,7 @@
 !/apps/files
 !/apps/federation
 !/apps/federatedfilesharing
+!/apps/sharebymail
```
please alphabetic order

schiessle replied:
done

nickvergessen commented on apps/sharebymail/appinfo/app.php:
```diff
@@ -0,0 +1,22 @@
+ *
+ */
+
+$app = new \OCP\AppFramework\App('sharebymail');
```
Is this file really needed?

schiessle replied:
no... I removed it

nickvergessen commented on core/shipped.json:
```diff
@@ -34,7 +34,8 @@
     "user_saml",
-    "workflowengine"
+    "workflowengine",
+    "sharebymail"
```
alphabetic order?

schiessle replied:
done

schiessle commented:
Please ignore the failing tests... They are not related to the actual code

@MorrisJobke any idea about this issue for example https://drone.nextcloud.com/nextcloud/server/645/21 can we do anything to make sure that tests don't run out of space?

Please review this... Would be great to get this in soon so that we can move forward. cc @LukasReschke @rullzer @nickvergessen @icewind1991

MorrisJobke commented:
> @MorrisJobke any idea about this issue for example https://drone.nextcloud.com/nextcloud/server/645/21 can we do anything to make sure that tests don't run out of space?

They actually do, because we have one server with very little space - I rearranged some stuff this evening. I restarted the failed tests.

LukasReschke reviewed (changes requested):
👎  until the new files have proper tests in place.

See https://codecov.io/gh/nextcloud/server/compare/f2cae3cee127220a5fde2726af8074837ee81ed1...9dc71f3033a1e2e95cdbec94e14ca05b54a8d800

rullzer commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -196,6 +199,28 @@ protected function formatShare(\OCP\Share\IShare $share, Node $recipientNode = n
+	 */
+	private function getDisplayNameFromAddressBook($query, $property) {
+		// FIXME: If we inject the contacts manager it gets initialized bofore any address books are registered
+		$result = \OC::$server->getContactsManager()->search($query, [$property]);
```
Mmm we should find a nice solution for that. But for now I guess it is OK

rullzer commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -400,6 +425,17 @@ public function createShare(
 			$share->setSharedWith($shareWith);
 			$share->setPermissions($permissions);
+		} else if ($shareType === \OCP\Share::SHARE_TYPE_EMAIL) {
+			if ($share->getNodeType() === 'file') {
```
Note for myself to change this for https://github.com/nextcloud/server/pull/1836

rullzer commented on apps/files_sharing/lib/Controller/ShareesAPIController.php:
```diff
@@ -584,6 +545,71 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
+	 * @param string $search
+	 * @return array
+	 */
+	protected function getEmail($search) {
```
Why is this sigular

rullzer commented on apps/files_sharing/lib/Controller/ShareesAPIController.php:
```diff
@@ -584,6 +545,71 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
+			];
+		}
+
+		$this->reachedEndFor[] = 'emails';
```
But this not

nickvergessen replied:
yeah, this needs to be the key of array that is being sent.

rullzer commented on core/js/sharedialogview.js:
```diff
@@ -154,10 +154,16 @@
 						var users   = result.ocs.data.exact.users.concat(result.ocs.data.users);
 						var groups  = result.ocs.data.exact.groups.concat(result.ocs.data.groups);
 						var remotes = result.ocs.data.exact.remotes.concat(result.ocs.data.remotes);
+						if (typeof(result.ocs.data.emails) !== 'undefined') {
```
can this really be undefined?

schiessle replied:
it happened to me during testing

nickvergessen replied:
Then you didn't initialize something correctly...

nickvergessen replied:
Then you can also kill this

rullzer commented on core/js/sharedialogview.js:
```diff
@@ -397,24 +408,24 @@
 
 		_renderSharePlaceholderPart: function () {
-			var sharePlaceholder = t('core', 'Share with users…');
+			var sharePlaceholder = t('core', 'Share with users, or by mail...');
```
We need to know in js if this is allowed. If the app is not enabled you can't share by mail.

rullzer commented:
Nice work @schiessle 
Probabaly needs some polishing afterwards but lets take care of that later.

Link was send sucsessfully here. But I'll test in a bit more detail.

Also needs a rebase.

schiessle commented:
> :-1: until the new files have proper tests in place.

@LukasReschke tests are coming in... don't panic :wink: You can still already review and test the code/functionality.

nickvergessen commented on apps/files_sharing/lib/Controller/ShareesAPIController.php:
```diff
@@ -560,13 +504,30 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
+			$this->result['exact']['emails'] = $mailResults['exact'];
+		} else if (!$mailResults['exactIdMatch'] && $remoteResults['exactIdMatch']) {
+			$this->result['remotes'] = $remoteResults['results'];
+			$this->result['exact']['remotes'] = $remoteResults['exact'];
```
please define the empty emails array here as well. We always send all keys.

schiessle replied:
they are already defined, see: https://github.com/nextcloud/server/blob/9dc71f3033a1e2e95cdbec94e14ca05b54a8d800/apps/files_sharing/lib/Controller/ShareesAPIController.php#L81

nickvergessen commented on apps/files_sharing/lib/Controller/ShareesAPIController.php:
```diff
@@ -560,13 +504,30 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
+		// that the exact same email address and federated cloud id exists
+		if ($mailResults['exactIdMatch'] && !$remoteResults['exactIdMatch']) {
+			$this->result['emails'] = $mailResults['results'];
+			$this->result['exact']['emails'] = $mailResults['exact'];
```
please define the empty remotes array here as well. We always send all keys.

schiessle replied:
they are already defined, see: https://github.com/nextcloud/server/blob/9dc71f3033a1e2e95cdbec94e14ca05b54a8d800/apps/files_sharing/lib/Controller/ShareesAPIController.php#L81

schiessle committed:
always show correct place holder, mention share by mail only if the share provider is enabled

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
add tests for the share-by-mail provider

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

rullzer commented:
We need to adjust this provider also for https://github.com/nextcloud/server/pull/339 Depending on which one goes in first.

schiessle committed:
add new method getSharesInFolder(), see #339

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
remove unused variable

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
import exception

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle commented:
> We need to adjust this provider also for #339 Depending on which one goes in first.

Done. I also added a lot of unit tests (cc @LukasReschke)... Final review round :wink:

rullzer commented:
I'd say lets get this in :+1:

LukasReschke commented:
LGTM

LukasReschke commented:
@nickvergessen request changes are pending here.

nickvergessen commented:
Yeah lets do that...
